Flaky test: test_dictionaries_all_layouts_separate_sources/test_clickhouse_local.py::test_ranged

_Important: This issue was automatically generated and is used by CI for matching failures. DO NOT modify the body content. DO NOT remove labels._

Test name: test_dictionaries_all_layouts_separate_sources/test_clickhouse_local.py::test_ranged
Failure reason: heap-use-after-free
CI report: [Integration tests (amd_tsan, 3/6)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=92511&sha=7c977e1b8d909f9850b379a7e5d330e3716a5905&name_0=PR&name_1=Integration%20tests%20%28amd_tsan%2C%203%2F6%29)

CIDB statistics: [cidb](None)

Test output:
```
File: test_dictionaries_all_layouts_separate_sources/test_clickhouse_local.py:70 - in started_cluster
    cluster.shutdown()
File: helpers/cluster.py:4006 - in shutdown
    raise Exception(
E   Exception: Sanitizer assert found for instance ==================
E   WARNING: ThreadSanitizer: heap-use-after-free (pid=8)
E     Read of size 8 at 0x7284000321b8 by thread T678:
E       #0 std::__1::weak_ptr<DB::Context>::lock() const ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:1374:18 (clickhouse+0x1dfab9df) (BuildId: ad9bf0aeae8e013116ff60e682b0cf5bbd00c0ff)
E       #1 DB::WithContextImpl<std::__1::shared_ptr<DB::Context>>::getContext() const ci/tmp/build/./src/Interpreters/Context_fwd.h:41:28 (clickhouse+0x1dfab9df)
E       #2 DB::IMergeTreeDataPart::clearCaches() ci/tmp/build/./src/Storages/MergeTree/IMergeTreeDataPart.cpp:711:34 (clickhouse+0x2592baad) (BuildId: ad9bf0aeae8e013116ff60e682b0cf5bbd00c0ff)
E       #3 DB::IMergeTreeDataPart::removeIfNeeded() ci/tmp/build/./src/Storages/MergeTree/IMergeTreeDataPart.cpp:745:9 (clickhouse+0x2592c0d3) (BuildId: ad9bf0aeae8e013116ff60e682b0cf5bbd00c0ff)
E       #4 DB::MergeTreeDataPartCompact::~MergeTreeDataPartCompact() ci/tmp/build/./src/Storages/MergeTree/MergeTreeDataPartCompact.cpp:289:9 (clickhouse+0x25bc4bb7) (BuildId: ad9bf0aeae8e013116ff60e682b0cf5bbd00c0ff)
E       #5 void std::__1::__destroy_at[abi:ne210105]<DB::MergeTreeDataPartCompact, 0>(DB::MergeTreeDataPartCompact*) ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/construct_at.h:61:11 (clickhouse+0x25bb1f68) (BuildId: ad9bf0aeae8e013116ff60e682b0cf5bbd00c0ff)
E       #6 void std::__1::allocator_traits<std::__1::allocator<DB::MergeTreeDataPartCompact>>::destroy[abi:ne210105]<DB::MergeTreeDataPartCompact, 0>(std::__1::allocator<DB::MergeTreeDataPartCompact>&, DB::MergeTreeDataPartCompact*) ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/allocator_traits.h:313:5 (clickhouse+0x25bb1f68)
E       #7 void std::__1::__shared_ptr_emplace<DB::MergeTreeDataPartCompact, std::__1::allocator<DB::MergeTreeDataPartCompact>>::__on_zero_shared_impl[abi:ne210105]<std::__1::allocator<DB::MergeTreeDataPartCompact>, 0>() ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:181:5 (clickhouse+0x25bb1f68)
E       #8 std::__1::__shared_ptr_emplace<DB::MergeTreeDataPartCompact, std::__1::allocator<DB::MergeTreeDataPartCompact>>::__on_zero_shared() ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:184:78 (clickhouse+0x25bb1f68)
E       #9 std::__1::__shared_count::__release_shared[abi:ne210105]() ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/shared_count.h:92:7 (clickhouse+0x202caf3a) (BuildId: ad9bf0aeae8e013116ff60e682b0cf5bbd00c0ff)
E       #10 std::__1::__shared_weak_count::__release_shared[abi:ne210105]() ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/shared_count.h:121:25 (clickhouse+0x202caf3a)
E       #11 std::__1::shared_ptr<DB::IMergeTreeDataPart const>::~shared_ptr[abi:ne210105]() ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:561:17 (clickhouse+0x202caf3a)
E       #12 DB::RangesInDataPart::~RangesInDataPart() ci/tmp/build/./src/Storages/MergeTree/RangesInDataPart.h:87:8 (clickhouse+0x202caf3a)
E       #13 void std::__1::__destroy_at[abi:ne210105]<DB::RangesInDataPart, 0>(DB::RangesInDataPart*) ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/construct_at.h:61:11 (clickhouse+0x25e081ee) (BuildId: ad9bf0aeae8e013116ff60e682b0cf5bbd00c0ff)
E       #14 void std::__1::allocator_traits<std::__1::allocator<DB::RangesInDataPart>>::destroy[abi:ne210105]<DB::RangesInDataPart, 0>(std::__1::allocator<DB::RangesInDataPart>&, DB::RangesInDataPart*) ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/allocator_traits.h:313:5 (clickhouse+0x25e081ee)
E       #15 std::__1::vector<DB::RangesInDataPart, std::__1::allocator<DB::RangesInDataPart>>::__base_destruct_at_end[abi:ne210105](DB::RangesInDataPart*) ci/tmp/build/./contrib/llvm-project/libcxx/include/__vector/vector.h:763:7 (clickhouse+0x25e081ee)
E       #16 std::__1::vector<DB::RangesInDataPart, std::__1::allocator<DB::RangesInDataPart>>::clear[abi:ne210105]() ci/tmp/build/./contrib/llvm-project/libcxx/include/__vector/vector.h:543:5 (clickhouse+0x25e081ee)
E       #17 std::__1::vector<DB::RangesInDataPart, std::__1::allocator<DB::RangesInDataPart>>::__destroy_vector::operator()[abi:ne210105]() ci/tmp/build/./contrib/llvm-project/libcxx/include/__vector/vector.h:251:16 (clickhouse+0x25e081ee)
E       #18 std::__1::vector<DB::RangesInDataPart, std::__1::allocator<DB::RangesInDataPart>>::~vector[abi:ne210105]() ci/tmp/build/./contrib/llvm-project/libcxx/include/__vector/vector.h:262:67 (clickhouse+0x25e081ee)
E       #19 DB::MergeTreeReadPoolBase::~MergeTreeReadPoolBase() ci/tmp/build/./src/Storages/MergeTree/MergeTreeReadPoolBase.h:14:7 (clickhouse+0x25e081ee)
E       #20 DB::MergeTreeReadPoolInOrder::~MergeTreeReadPoolInOrder() ci/tmp/build/./src/Storages/MergeTree/MergeTreeReadPoolInOrder.h:7:7 (clickhouse+0x2715980e) (BuildId: ad9bf0aeae8e013116ff60e682b0cf5bbd00c0ff)
E       #21 void std::__1::__destroy_at[abi:ne210105]<DB::MergeTreeReadPoolInOrder, 0>(DB::MergeTreeReadPoolInOrder*) ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/construct_at.h:61:11 (clickhouse+0x271055e7) (BuildId: ad9bf0aeae8e013116ff60e682b0cf5bbd00c0ff)
E       #22 void std::__1::allocator_traits<std::__1::allocator<DB::MergeTreeReadPoolInOrder>>::destroy[abi:ne210105]<DB::MergeTreeReadPoolInOrder, 0>(std::__1::allocator<DB::MergeTreeReadPoolInOrder>&, DB::MergeTreeReadPoolInOrder*) ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/allocator_traits.h:313:5 (clickhouse+0x271055e7)
E       #23 void std::__1::__shared_ptr_emplace<DB::MergeTreeReadPoolInOrder, std::__1::allocator<DB::MergeTreeReadPoolInOrder>>::__on_zero_shared_impl[abi:ne210105]<std::__1::allocator<DB::MergeTreeReadPoolInOrder>, 0>() ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:181:5 (clickhouse+0x271055e7)
E       #24 std::__1::__shared_ptr_emplace<DB::MergeTreeReadPoolInOrder, std::__1::allocator<DB::MergeTreeReadPoolInOrder>>::__on_zero_shared() ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:184:78 (clickhouse+0x271055e7)
E       #25 std::__1::__shared_count::__release_shared[abi:ne210105]() ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/shared_count.h:92:7 (clickhouse+0x270f9eff) (BuildId: ad9bf0aeae8e013116ff60e682b0cf5bbd00c0ff)
E       #26 std::__1::__shared_weak_count::__release_shared[abi:ne210105]() ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/shared_count.h:121:25 (clickhouse+0x270f9eff)
E       #27 std::__1::shared_ptr<DB::IMergeTreeReadPool>::~shared_ptr[abi:ne210105]() ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:561:17 (clickhouse+0x270f9eff)
E       #28 DB::MergeTreeSelectProcessor::~MergeTreeSelectProcessor() ci/tmp/build/./src/Storages/MergeTree/MergeTreeSelectProcessor.h:99:7 (clickhouse+0x270f9eff)
E       #29 std::__1::default_delete<DB::MergeTreeSelectProcessor>::operator()[abi:ne210105](DB::MergeTreeSelectProcessor*) const ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:77:5 (clickhouse+0x2711163e) (BuildId: ad9bf0aeae8e013116ff60e682b0cf5bbd00c0ff)
E       #30 std::__1::unique_ptr<DB::MergeTreeSelectProcessor, std::__1::default_delete<DB::MergeTreeSelectProcessor>>::reset[abi:ne210105](DB::MergeTreeSelectProcessor*) ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:290:7 (clickhouse+0x2711163e)
E       #31 std::__1::unique_ptr<DB::MergeTreeSelectProcessor, std::__1::default_delete<DB::MergeTreeSelectProcessor>>::~unique_ptr[abi:ne210105]() ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:259:71 (clickhouse+0x2711163e)
E       #32 DB::MergeTreeSource::~MergeTreeSource() ci/tmp/build/./src/Storages/MergeTree/MergeTreeSource.cpp:147:35 (clickhouse+0x2711163e)
E       #33 void std::__1::__destroy_at[abi:ne210105]<DB::MergeTreeSource, 0>(DB::MergeTreeSource*) ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/construct_at.h:61:11 (clickhouse+0x2710469d) (BuildId: ad9bf0aeae8e013116ff60e682b0cf5bbd00c0ff)
E       #34 void std::__1::allocator_traits<std::__1::allocator<DB::MergeTreeSource>>::destroy[abi:ne210105]<DB::MergeTreeSource, 0>(std::__1::allocator<DB::MergeTreeSource>&, DB::MergeTreeSource*) ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/allocator_traits.h:313:5 (clickhouse+0x2710469d)
E       #35 void std::__1::__shared_ptr_emplace<DB::MergeTreeSource, std::__1::allocator<DB::MergeTreeSource>>::__on_zero_shared_impl[abi:ne210105]<std::__1::allocator<DB::MergeTreeSource>, 0>() ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:181:5 (clickhouse+0x2710469d)
E       #36 std::__1::__shared_ptr_emplace<DB::MergeTreeSource, std::__1::allocator<DB::MergeTreeSource>>::__on_zero_shared() ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:184:78 (clickhouse+0x2710469d)
E       #37 std::__1::__shared_count::__release_shared[abi:ne210105]() ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/shared_count.h:92:7 (clickhouse+0x1e5210f0) (BuildId: ad9bf0aeae8e013116ff60e682b0cf5bbd00c0ff)
E       #38 std::__1::__shared_weak_count::__release_shared[abi:ne210105]() ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/shared_count.h:121:25 (clickhouse+0x1e5210f0)
E       #39 std::__1::shared_ptr<DB::IProcessor>::~shared_ptr[abi:ne210105]() ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:561:17 (clickhouse+0x1e5210f0)
E       #40 void std::__1::__destroy_at[abi:ne210105]<std::__1::shared_ptr<DB::IProcessor>, 0>(std::__1::shared_ptr<DB::IProcessor>*) ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/construct_at.h:61:11 (clickhouse+0x1e5210f0)
E       #41 void std::__1::allocator_traits<std::__1::allocator<std::__1::shared_ptr<DB::IProcessor>>>::destroy[abi:ne210105]<std::__1::shared_ptr<DB::IProcessor>, 0>(std::__1::allocator<std::__1::shared_ptr<DB::IProcessor>>&, std::__1::shared_ptr<DB::IProcessor>*) ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/allocator_traits.h:313:5 (clickhouse+0x1e5210f0)
E       #42 std::__1::vector<std::__1::shared_ptr<DB::IProcessor>, std::__1::allocator<std::__1::shared_ptr<DB::IProcessor>>>::__base_destruct_at_end[abi:ne210105](std::__1::shared_ptr<DB::IProcessor>*) ci/tmp/build/./contrib/llvm-project/libcxx/include/__vector/vector.h:763:7 (clickhouse+0x1e5210f0)
~~~~~ truncated 179 lines ~~~~~
```

--- a/src/Dictionaries/CacheDictionary.cpp
+++ b/src/Dictionaries/CacheDictionary.cpp
@@ -646,11 +646,11 @@ void CacheDictionary<dictionary_key_type>::update(CacheDictionaryUpdateUnitPtr<d
 
             Columns fetched_columns_during_update = fetch_request.makeAttributesResultColumnsNonMutable();
 
-            DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
-            io.pipeline.setConcurrencyControl(false);
-
             io.executeWithCallbacks([&]()
             {
+                DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
+                io.pipeline.setConcurrencyControl(false);
+
                 Block block;
                 while (executor.pull(block))
                 {

--- a/src/Dictionaries/DictionaryHelpers.h
+++ b/src/Dictionaries/DictionaryHelpers.h
@@ -594,10 +594,10 @@ Block mergeBlockWithPipe(
 
     auto result_fetched_columns = block_to_update.cloneEmptyColumns();
 
-    PullingPipelineExecutor executor(io.pipeline);
-
     io.executeWithCallbacks([&]()
     {
+        PullingPipelineExecutor executor(io.pipeline);
+
         Block block;
         while (executor.pull(block))
         {

--- a/src/Dictionaries/DirectDictionary.cpp
+++ b/src/Dictionaries/DirectDictionary.cpp
@@ -90,15 +90,15 @@ Columns DirectDictionary<dictionary_key_type>::getColumns(
 
     BlockIO io = loadKeys(requested_keys, key_columns);
 
-    QueryPipeline pipeline(getSourcePipe(io.pipeline, use_async_executor));
-    PullingPipelineExecutor executor(pipeline);
-
     Stopwatch watch;
     size_t block_num = 0;
     size_t rows_num = 0;
 
     io.executeWithCallbacks([&]()
     {
+        QueryPipeline pipeline(getSourcePipe(io.pipeline, use_async_executor));
+        PullingPipelineExecutor executor(pipeline);
+
         Block block;
         while (executor.pull(block))
         {
@@ -262,13 +262,13 @@ ColumnUInt8::Ptr DirectDictionary<dictionary_key_type>::hasKeys(
 
     BlockIO io = loadKeys(requested_keys, key_columns);
 
-    QueryPipeline pipeline(getSourcePipe(io.pipeline, use_async_executor));
-    PullingPipelineExecutor executor(pipeline);
-
     size_t keys_found = 0;
 
     io.executeWithCallbacks([&]()
     {
+        QueryPipeline pipeline(getSourcePipe(io.pipeline, use_async_executor));
+        PullingPipelineExecutor executor(pipeline);
+
         Block block;
         while (executor.pull(block))
         {

--- a/src/Dictionaries/FlatDictionary.cpp
+++ b/src/Dictionaries/FlatDictionary.cpp
@@ -456,12 +456,13 @@ void FlatDictionary::updateData()
 
     if (!update_field_loaded_block || update_field_loaded_block->rows() == 0)
     {
-        DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
-        io.pipeline.setConcurrencyControl(false);
         update_field_loaded_block.reset();
 
         io.executeWithCallbacks([&]()
         {
+            DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
+            io.pipeline.setConcurrencyControl(false);
+
             Block block;
             while (executor.pull(block))
             {
@@ -501,11 +502,11 @@ void FlatDictionary::loadData()
     {
         BlockIO io = source_ptr->loadAll();
 
-        DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
-        io.pipeline.setConcurrencyControl(false);
-
         io.executeWithCallbacks([&]()
         {
+            DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
+            io.pipeline.setConcurrencyControl(false);
+
             Block block;
             while (executor.pull(block))
                 blockToAttributes(block);

--- a/src/Dictionaries/HashedArrayDictionary.cpp
+++ b/src/Dictionaries/HashedArrayDictionary.cpp
@@ -483,12 +483,13 @@ void HashedArrayDictionary<dictionary_key_type, sharded>::updateData()
 
     if (!update_field_loaded_block || update_field_loaded_block->rows() == 0)
     {
-        DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
-        io.pipeline.setConcurrencyControl(false);
         update_field_loaded_block.reset();
 
         io.executeWithCallbacks([&]()
         {
+            DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
+            io.pipeline.setConcurrencyControl(false);
+
             Block block;
             while (executor.pull(block))
             {
@@ -985,9 +986,6 @@ void HashedArrayDictionary<dictionary_key_type, sharded>::loadData()
 
         BlockIO io = source_ptr->loadAll();
 
-        DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
-        io.pipeline.setConcurrencyControl(false);
-
         UInt64 pull_time_microseconds = 0;
         UInt64 process_time_microseconds = 0;
 
@@ -997,6 +995,9 @@ void HashedArrayDictionary<dictionary_key_type, sharded>::loadData()
 
         io.executeWithCallbacks([&]()
         {
+            DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
+            io.pipeline.setConcurrencyControl(false);
+
             Block block;
             while (true)
             {

--- a/src/Dictionaries/HashedDictionary.h
+++ b/src/Dictionaries/HashedDictionary.h
@@ -879,12 +879,13 @@ void HashedDictionary<dictionary_key_type, sparse, sharded>::updateData()
 
     if (!update_field_loaded_block || update_field_loaded_block->rows() == 0)
     {
-        DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
-        io.pipeline.setConcurrencyControl(false);
         update_field_loaded_block.reset();
 
         io.executeWithCallbacks([&]()
         {
+            DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
+            io.pipeline.setConcurrencyControl(false);
+
             Block block;
             while (executor.pull(block))
             {
@@ -1161,11 +1162,12 @@ void HashedDictionary<dictionary_key_type, sparse, sharded>::loadData()
 
         BlockIO io = source_ptr->loadAll();
 
-        DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
-        io.pipeline.setConcurrencyControl(false);
         DictionaryKeysArenaHolder<dictionary_key_type> arena_holder;
         io.executeWithCallbacks([&]()
         {
+            DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
+            io.pipeline.setConcurrencyControl(false);
+
             Block block;
             while (executor.pull(block))
             {

--- a/src/Dictionaries/IPAddressDictionary.cpp
+++ b/src/Dictionaries/IPAddressDictionary.cpp
@@ -415,11 +415,11 @@ void IPAddressDictionary::loadData()
     {
         BlockIO io = source_ptr->loadAll();
 
-        DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
-        io.pipeline.setConcurrencyControl(false);
-
         io.executeWithCallbacks([&]()
         {
+            DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
+            io.pipeline.setConcurrencyControl(false);
+
             Block block;
             while (executor.pull(block))
             {

--- a/src/Dictionaries/PolygonDictionary.cpp
+++ b/src/Dictionaries/PolygonDictionary.cpp
@@ -291,10 +291,11 @@ void IPolygonDictionary::loadData()
 {
     BlockIO io = source_ptr->loadAll();
 
-    DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
-    io.pipeline.setConcurrencyControl(false);
     io.executeWithCallbacks([&]()
     {
+        DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
+        io.pipeline.setConcurrencyControl(false);
+
         Block block;
         while (executor.pull(block))
             blockToAttributes(block);

--- a/src/Dictionaries/RangeHashedDictionary.h
+++ b/src/Dictionaries/RangeHashedDictionary.h
@@ -547,10 +547,11 @@ void RangeHashedDictionary<dictionary_key_type>::loadData()
     {
         BlockIO io = source_ptr->loadAll();
 
-        DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
-        io.pipeline.setConcurrencyControl(false);
         io.executeWithCallbacks([&]()
         {
+            DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
+            io.pipeline.setConcurrencyControl(false);
+
             Block block;
             while (executor.pull(block))
             {
@@ -702,12 +703,13 @@ void RangeHashedDictionary<dictionary_key_type>::updateData()
 
     if (!update_field_loaded_block || update_field_loaded_block->rows() == 0)
     {
-        DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
-        io.pipeline.setConcurrencyControl(false);
         update_field_loaded_block.reset();
 
         io.executeWithCallbacks([&]()
         {
+            DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
+            io.pipeline.setConcurrencyControl(false);
+
             Block block;
             while (executor.pull(block))
             {

--- a/src/Dictionaries/RegExpTreeDictionary.cpp
+++ b/src/Dictionaries/RegExpTreeDictionary.cpp
@@ -318,11 +318,11 @@ void RegExpTreeDictionary::loadData()
     {
         BlockIO io = source_ptr->loadAll();
 
-        DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
-        io.pipeline.setConcurrencyControl(false);
-
         io.executeWithCallbacks([&]()
         {
+            DictionaryPipelineExecutor executor(io.pipeline, configuration.use_async_executor);
+            io.pipeline.setConcurrencyControl(false);
+
             Block block;
             while (executor.pull(block))
             {

